lib/rspec/matchers/built_in/compound.rb

```diff
@@ -97,11 +99,19 @@ def compound_failure_message
         end
 
         def matcher_1_matches?
-          evaluator.matcher_matches?(matcher_1)
+          if @matcher_1_matches.nil?
+            @matcher_1_matches = evaluator.matcher_matches?(matcher_1)
+          else
+            @matcher_1_matches
+          end
```

Comment:
To appease RuboCop's `Metrics/ClassLength` it's possible to write this as a guard statement:
```ruby
return @matcher_1_matches unless @matcher_1_matches.nil?
@matcher_1_matches = evaluator.matcher_matches?(matcher_1)
```